components/command/runner.go

```diff
@@ -21,48 +22,113 @@ type Args struct {
-		Update:     osCommand.BuildCommandString(args.Update, args.Environment, args.Sudo, args.RequirePasswordFromStdin, config.user),
-		Delete:     osCommand.BuildCommandString(args.Delete, args.Environment, args.Sudo, args.RequirePasswordFromStdin, config.user),
-		Triggers:   args.Triggers,
-		Stdin:      args.Stdin,
+func (args *Args) toRemoteCommandArgs(config RunnerConfiguration, osCommand OSCommand) (*remote.CommandArgs, error) {
+	// Ensure no local arguments are passed to remote commands
+	if args.LocalAssetPaths != nil {
+		return nil, fmt.Errorf("local asset paths are not supported in remote commands")
+	}
+	if args.LocalDir != nil {
+		return nil, fmt.Errorf("local dir is not supported in remote commands")
 	}
+
+	return &remote.CommandArgs{
+		Connection:  config.connection,
+		Create:      osCommand.BuildCommandString(args.Create, args.Environment, args.Sudo, args.RequirePasswordFromStdin, config.user),
+		Update:      osCommand.BuildCommandString(args.Update, args.Environment, args.Sudo, args.RequirePasswordFromStdin, config.user),
+		Delete:      osCommand.BuildCommandString(args.Delete, args.Environment, args.Sudo, args.RequirePasswordFromStdin, config.user),
+		Environment: args.Environment,
+		Triggers:    args.Triggers,
+		Stdin:       args.Stdin,
+	}, nil
 }
 
 // Transformer is a function that can be used to modify the command name and args.
 // Examples: swapping `args.Delete` with `args.Create`, or adding `args.Triggers`, or editing the name
 type Transformer func(name string, args Args) (string, Args)
 
-type runnerConfiguration struct {
+type RunnerConfiguration struct {
 	user       string
 	connection remote.ConnectionInput
 }
 
-type Runner struct {
+type Command interface {
+	pulumi.Resource
+
+	StdoutOutput() pulumi.StringOutput
+	StderrOutput() pulumi.StringOutput
+}
+
+type LocalCommand struct {
+	*local.Command
+}
+
+type RemoteCommand struct {
+	*remote.Command
+}
+
+var _ Command = &RemoteCommand{}
+var _ Command = &LocalCommand{}
+
+func (c *LocalCommand) StdoutOutput() pulumi.StringOutput {
+	return c.Command.Stdout
+}
+
+func (c *LocalCommand) StderrOutput() pulumi.StringOutput {
+	return c.Command.Stderr
+}
+
+func (c *RemoteCommand) StdoutOutput() pulumi.StringOutput {
+	return c.Command.Stdout
+}
+
+func (c *RemoteCommand) StderrOutput() pulumi.StringOutput {
+	return c.Command.Stderr
+}
+
+type Runner interface {
+	Environment() config.Env
+	Namer() namer.Namer
+	Config() RunnerConfiguration
+	OsCommand() OSCommand
+
+	Command(name string, args *Args, opts ...pulumi.ResourceOption) (Command, error)
+	NewCopyFile(name string, localPath, remotePath pulumi.StringInput, opts ...pulumi.ResourceOption) (pulumi.Resource, error)
+	CopyWindowsFile(name string, src, dst pulumi.StringInput, opts ...pulumi.ResourceOption) (pulumi.Resource, error)
+	CopyUnixFile(name string, src, dst pulumi.StringInput, opts ...pulumi.ResourceOption) (pulumi.Resource, error)
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
Should be possible to make this not exported, and allow only a certain package to use them, aka the filemanager